[dnm] kvserver: subject MsgApp to admission control

tbg opened the pull request:
Context: https://github.com/cockroachdb/cockroach/issues/79215#issuecomment-1126093924

- [ ] block on raft receive queue (first commit)
- [ ] block in handleRaftReady (second commit)
- [ ] treat followers with high r-amp as probing (TBD)
- [ ] delay ready handling until admission control passed (TBD)

Release note: None

tbg committed:
ui: use metric name as custom chart title

Custom charts were previously difficult to read because the graph didn't
include any indication of what was being graphed. This is now rectified
by printing the metrics name. This isn't the most user friendly thing
(ideally we would also include the actual title, and make the help text
available) but with my limited TSX knowledge this is solving 90% of the
problem with a 61 character change.

Related to https://github.com/cockroachdb/cockroach/issues/81035.

Release note: None

tbg committed:
[dnm] kvserver: subject MsgApp to admission control

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81289)
<!-- Reviewable:end -->

tbg commented:
First experiment: Naively backpressuring on the incoming raft stream via local admission control in https://github.com/cockroachdb/cockroach/blob/1f46dd08d8ad3fa82038abb6852a5920f0f93305/pkg/kv/kvserver/store_raft.go#L299-L301. This wreaked havoc immediately (including unavailable liveness, I think) since the liveness leaseholder was probably on n3, and n3 was backpressuring on each incoming MsgApp regardless of which range it was for. "Most" quota pools in the system were depleted:

<img width="935" alt="image" src="https://user-images.githubusercontent.com/5076964/168579967-58bae2d3-0560-49d1-aa67-04dee331c6c7.png">

It would've been interesting to see the metrics in that state but alas, metrics timeseries ranges were also all unavailable.

tbg committed:
kvserver: apply admission control in handleRaftReady

Wait in handleRaftReady, which isn't great either, but at least allows
us to target individual ranges (mod Go scheduler concurrency).

Release note: None

tbg committed:
roachprod: improve store attributes

They would previously be `store1` since we mostly run single-store
deployments. It's often helpful to target a particular node, for
example for lease preferences.

Before:

```
 select node_id, attrs from crdb_internal.kv_store_status;
  node_id |               attrs
----------+-------------------------------------
        1 | ["store1"]
```

After:

```
  node_id |               attrs
        1 | ["node1", "node1store1", "store1"]
```

Release note: None

tbg committed:
[dnm] helper to set up prometheus

<what was there before: Previously, ...>
<why it needed to change: This was inadequate because ...>
<what you did about it: To address this, this patch ...>

Release note (<category, see below>): <what> <show> <why>

tbg committed:
kvserver: add `raft.rcvd.admission_wait` metric

Sloppy for now

Release note: None

tbg commented:
For a second experiment, I moved the backpressure to `handleRaftReadyRaftMuLocked`. This has the advantage that it allows filtering by RangeID and that it doesn't block the entire receive queue to the store indiscriminately. The disadvantage is that this is below-raft throttling, so we're now holding `raftMu` for extended periods of time. I also pre-split 1000 ranges for this workload. Unfortunately, n1 (which has all leases pinned to it) was starting to get high read-amp as well, but we can see in the graphs that for quite some time before that, n3 is struggling with high read-amp but that it is staying bounded - meaning admission control on the raft follower path is doing its job:

![image](https://user-images.githubusercontent.com/5076964/168657323-607094a8-e723-4959-be82-eddd0a08934a.png)

tbg committed:
kvserver: limit admission control to s3

In the experiment, n1 ended up with high r-amp and started
backpressuring below raft as well. This isn't the intention
since those writes will be doubly backpressured. We ended up
in a regime where n1 is constantly overloaded and n3 is actually
doing ok, perhaps metaunstable.

Release note: None

tbg commented:
There's an internal thread [here](https://cockroachlabs.slack.com/archives/CAC6K3SLU/p1652782655825589?thread_ts=1652728976.562359&cid=CAC6K3SLU) about whether this experiment is flawed in the sense that the cluster cannot possibly sustain 6mb/s, since I'm running it with the stock binary now (leases on n1 and n2) and n1 and n2 are the ones that have high read amp, despite being provisioned at 250mb/s (double the default).

tbg commented:
This is being productionized in #83851.
